imap/jmap_util.c

```diff
@@ -696,3 +697,128 @@ EXPORTED const char *jmap_decode_to_utf8(const char *charset, int encoding,
     *val = text;
     return text;
 }
+
+EXPORTED const char *jmap_encode_rawdata_blobid(const char prefix,
+                                                const char *mboxid,
+                                                uint32_t uid,
+                                                const char *userid,
+                                                const char *subpart,
+                                                struct message_guid *guid,
+                                                struct buf *dst)
+{
+    buf_reset(dst);
+
+    /* Set smart blob prefix */
+    buf_putc(dst, prefix);
+
+    /* Encode mailbox id */
+    buf_appendcstr(dst, mboxid);
+    
+    /* Encode message UID */
+    buf_printf(dst, "_%u", uid);
+
+    /* Encode user id */
+    if (userid) {
+        buf_putc(dst, '_');
+        char *b64userid = jmap_encode_base64_nopad(userid, strlen(userid));
+        if (!b64userid) {
+            buf_reset(dst);
+            return NULL;
+        }
+        buf_appendcstr(dst, b64userid);
+        free(b64userid);
+    }
+
+    /* Encode subpart */
+    if (subpart) {
+        buf_putc(dst, '_');
+        char *b64part = jmap_encode_base64_nopad(subpart, strlen(subpart));
+        if (!b64part) {
+            buf_reset(dst);
+            return NULL;
+        }
+        buf_appendcstr(dst, b64part);
+        free(b64part);
+
+        if (guid) {
+            /* Encode subpart data GUID */
+            buf_printf(dst, "_%s", message_guid_encode(guid));
+        }
+    }
+
+    return buf_cstring(dst);
+}
+
+EXPORTED int jmap_decode_rawdata_blobid(const char *blobid,
+                                        char **mboxidptr,
+                                        uint32_t *uidptr,
+                                        char **useridptr,
+                                        char **subpartptr,
+                                        struct message_guid *guid)
+{
+    char *mboxid = NULL;
+    uint32_t uid = 0;
+    char *userid = NULL;
+    char *subpart = NULL;
+    int is_valid = 0;
+
+    /* Decode mailbox id */
+    const char *base = blobid+1;
+    const char *p = strchr(base, '_');
+    if (!p) goto done;
+    mboxid = xstrndup(base, p-base);
+    if (!*mboxid) goto done;
+    base = p + 1;
+
+    /* Decode message UID */
+    if (*base == '\0') goto done;
+    char *endptr = NULL;
+    errno = 0;
+    uid = strtoul(base, &endptr, 10);
+    if (errno == ERANGE || (*endptr && *endptr != '_')) {
+        goto done;
+    }
+    base = endptr;
+
+    /* Decode userid */
+    if (useridptr && *base == '_') {
+        base += 1;
+        size_t len = strlen(base);
+        if (len) {
+            userid = jmap_decode_base64_nopad(base, len);
+            if (!userid) goto done;
+        }
+        base += len;
+    }
+
+    /* Decode subpart */
+    if (subpartptr && *base == '_') {
```

Comment:
Wouldn't this break if the caller passes a null useridptr and a non-null subpartptr (or the other way round) and the blobId contains both? I think that if the caller passes null arguments, then the decoder should validate but skip these parts in the blobId.

Reply:
I struggled with how best to handle this, since currently only CalendarEvent uses 'userid' and only Contact uses 'subpart' (which I renamed from 'property').  I thought about not having 'userid' argument and just passing the userid in as the 'subpart' because there currently isn't a use-case for 'userid' with Contact and there may never be one.  However, I could see a use-case for both 'userid' and 'subpart' with CalendarEvent, e.g., A per-user IMAGE property on a VEVENT.

Perhaps, we could write the encoder so that if a subpart is used, the userid MUST be encoded, even if empty, so mboxid__subpart_guid (consecutive underscores between mboxid and subpart).